src/core/ext/transport/chttp2/transport/hpack_parser.c

```diff
@@ -1650,6 +1650,7 @@ static void force_client_rst_stream(grpc_exec_ctx *exec_ctx, void *sp,
         &t->qbuf, grpc_chttp2_rst_stream_create(s->id, GRPC_HTTP2_NO_ERROR,
                                                 &s->stats.outgoing));
     grpc_chttp2_initiate_write(exec_ctx, t, "force_rst_stream");
+    grpc_chttp2_fail_pending_writes(exec_ctx, t, s, GRPC_ERROR_NONE);
```

Comment:
Will this not be invoked from mark_stream_closed?

Reply:
The main issue is not with fail_pending_writes, but with a race while sending trailing metadata, and actually deleting the stream from the stream_map. The stream_map size is currently used to check the number of streams in use currently. Earlier, the server would send the trailing metadata and only remove the stream after that, which gave rise to the possibility of the client sending another stream before the original was removed, creating the flake that we see.

The solution was to call mark_stream_closed before sending the data. To do this, I had to remove fail_pending_writes (from mark_stream_closed) which was causing a few issues.
(I actually got confused when I saw your comment. I thought maybe that I had left removing the line from mark_stream_closed, but then I realized that your link points to the current grpc master, and not the pull request :))

Reply:
Ahh oops! Looks like I was half reviewing your code, and half reviewing an old tab I had open at master. The dangers of too many tabs...

This is a great comment, I appreciate the detailed explanation.

Reply:
[accidentally wrote and delete comment. Re writing]

Oops, I was half reviewing your code, and half reviewing an old chrome tab open at master. The dangers of too many tabs...

Reply:
Actually, this just reminded me something. I changed the fix a bit and this was after noticing the fail_pending_writes. The fail_pending_writes change may no longer be required. The last I talked to Craig, as long as something is put on the qbuf, it should go out. Let me try adding the fail_pending_writes back in.